bundler/lib/bundler/checksum.rb

```diff
@@ -10,42 +75,70 @@ def initialize
         @store = {}
       end
 
-      def initialize_copy(o)
+      def initialize_copy(other)
         @store = {}
-        o.store.each do |k, v|
-          @store[k] = v.dup
+        other.store.each do |full_name, checksums|
+          store[full_name] = checksums.dup
         end
       end
 
-      def [](spec)
-        sums = @store[spec.full_name]
-
-        Checksum.new(spec.name, spec.version, spec.platform, sums&.transform_values(&:digest))
+      def checksums(full_name)
+        store[full_name]
       end
 
-      def register(spec, checksums)
-        register_full_name(spec.full_name, checksums)
+      def delete(full_name)
+        store.delete(full_name)
       end
 
-      def register_triple(name, version, platform, checksums)
-        register_full_name(GemHelpers.spec_full_name(name, version, platform), checksums)
+      def register_gem_from_file(full_name, file_source)
+        new_checksums = Checksum.from_file_source(file_source)
+        new_checksums.each do |checksum|
+          register full_name, checksum
+        end
+      rescue SecurityError
+        expected = checksums(full_name)
+
+        raise SecurityError, <<~MESSAGE
+          Bundler cannot continue installing #{full_name}.
+          The checksum for the downloaded `#{full_name}.gem` does not match \
+          the known checksum for the gem.
+          This means the contents of the downloaded \
+          gem is different from what was uploaded to the server \
+          or first used by your teammates, and could be a potential security issue.
+
+          To resolve this issue:
+          1. delete the downloaded gem located at: `#{file_source.path}`
+          2. run `bundle install`
+
+          If you are sure that the new checksum is correct, you can \
```

Comment:
I think we should not suggest to edit lockfile manually and also to disable checksum validation for whole project (Gemfile.lock).

Reply:
I brought this up in the [RFC](https://github.com/rubygems/rfcs/pull/50/files#diff-e042d284f6944e04fe33f6cd94cf0305182593c6e368c19ee039eaa094c64860R82) as an unknown. Since it was not originally added as part of this patch PR, we can solve it on the main PR.